Framework/Algorithms/src/CalculateAsymmetry.cpp

```diff
@@ -0,0 +1,240 @@
+          "Spectra size greater than the number of spectra!");
+    }
+
+    // check start and end times
+    double startX = getProperty("StartX");
+    double endX = getProperty("EndX");
+
+    if (startX > endX) {
+      g_log.warning()
+          << "Start time is after the end time. Swapping the start and end."
+          << '\n';
+      double tmp = endX;
+      endX = startX;
+      startX = tmp;
+    } else if (startX == endX) {
+      throw std::runtime_error("Start and end times are equal, there is no "
+                               "data to apply the algorithm to.");
+    }
+
+    auto xData = inputWS->histogram(specNum).binEdges();
+    if (startX < xData[0]) {
+      g_log.warning() << "Start time is before the first data point. Using "
+                         "first data point." << '\n';
+    }
+    if (endX > xData[xData.size() - 1]) {
+      g_log.warning()
+          << "End time is after the last data point. Using last data point."
+          << '\n';
+      g_log.warning() << "Data at late times may dominate the normalisation."
+                      << '\n';
+    }
+
+    const Mantid::API::Run &run = inputWS->run();
+    const double numGoodFrames = std::stod(run.getProperty("goodfrm")->value());
+    // inital estimate of N0
+    const double estNormConst = estimateNormalisationConst(
+        inputWS->histogram(specNum), numGoodFrames, startX, endX);
+    // Calculate the normalised counts
+    outputWS->setHistogram(
+        specNum, normaliseCounts(inputWS->histogram(specNum), numGoodFrames));
+    // get the normalisation constant
+    const double normConst =
+        getNormConstant(outputWS, spectra[i], estNormConst, startX, endX);
+    // calculate the asymmetry
+    outputWS->mutableY(specNum) /= normConst;
+    outputWS->mutableY(specNum) -= 1.0;
+    outputWS->mutableE(specNum) /= normConst;
+
+    prog.report();
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+
+  // Update Y axis units
+  outputWS->setYUnit("Asymmetry");
+
+  setProperty("OutputWorkspace", outputWS);
+}
+
+/**
+ * Calculate normalisation constant after the exponential decay has been removed
+ * to a linear fitting function
+ * @param ws ::  workspace
+ * @param wsIndex :: workspace index
+ * @param estNormConstant :: estimate of normalisation constant
+ * @param startX :: the smallest x value for the fit
+ * @param endX :: the largest x value for the fit
+ * @return normalisation constant
+*/
+double CalculateAsymmetry::getNormConstant(API::MatrixWorkspace_sptr ws,
+                                           int wsIndex,
+                                           const double estNormConstant,
+                                           const double startX,
+                                           const double endX) {
+  double retVal = 1.0;
+
+  API::IAlgorithm_sptr fit;
+  fit = createChildAlgorithm("Fit", -1, -1, true);
+  std::string tmpString = getProperty("FittingFunction");
```

Comment:
So I have had a play with the code and all of the fit algorithm has to stay within the loop. If I just move the createChildAlgorithm outside of the loop then the first spectrum is correct, but any after that are wrong. I think this is because it adds the new data to the fit function each time the normalization is called, without removing the previous data. In the current set up the fit function is automatically cleared as it goes out of scope between different spectra.